samples/scan/scan.cpp

```diff
@@ -57,104 +57,210 @@ struct identity<T, std::logical_and<T>> {
   static constexpr T value = true;
 };
 
+// Dummy struct to generate unique kernel name types
+template <typename T, typename U, size_t N>
+struct kernel_name {};
+
 /* Performs an inclusive scan with the given associative binary operation `Op`
- * in parallel using the provided accelerated hardware queue. Returns the result
- * as a vector of the same size as the input, and the input is unmodified.
+ * on the data in the `in` buffer. Runs in parallel on the provided accelerated
+ * hardware queue. Modifies the input buffer to contain the results of the scan.
  * Input size has to be a power of two. If the size isn't so, the input can
  * easily be padded to the nearest power of two with any values, and the scan on
  * the meaningful part of the data will stay the same. */
 template <typename T, typename Op>
-std::vector<T> par_scan(std::vector<T> const& in, sycl::queue& q) {
-  size_t size = in.size();
-  std::vector<T> out(size);
+void par_scan(sycl::buffer<T, 1>& in, sycl::queue& q) {
```

Comment:
Why the interface change - to avoid the costly copy?

Reply:
With the previous interface, the recursive invocation would have to make a host-side vector. This avoids that.

Reply:
Ah of course, it's recursive, I see now. Thanks!